backend/dataall/modules/loader.py

```diff
@@ -29,49 +31,199 @@ class ModuleInterface(ABC):
+    return modules
 
-def _import_module(name):
+
+def _load_module(name: str):
+    """
+    Loads a module but not initializing it
+    """
     try:
         importlib.import_module(f"{_MODULE_PREFIX}.{name}")
         return True
     except ModuleNotFoundError:
         return False
+
+
+def _initialize_modules(modes: List[ImportMode]):
+    """
+    Initialize all modules for supported modes. This method is using topological sorting for a graph of module
+    dependencies. It's needed to load module in a specific order: first modules to load are without dependencies.
+    It might help to avoid possible issues if there is a load in the module constructor (which can be the case
+    if a module supports a few importing modes).
+    """
+    modules = _all_modules()
+    dependencies = defaultdict(list)
+    degrees = defaultdict(int)
+    supported = []
+    for module in modules:
+        if module.is_supported(modes):
+            supported.append(module)
+            degrees[module] += len(module.depends_on())
+            for dependency in module.depends_on():
+                dependencies[dependency].append(module)
+
+    queue = deque()
+    for module in supported:
+        if degrees[module] == 0:
+            queue.append(module)
+
+    initialized = 0
+    while queue:
+        to_init = queue.popleft()
+        _initialize_module(to_init)
+        initialized += 1
+
+        for dependant in dependencies[to_init]:
+            degrees[dependant] -= 1
+            if degrees[dependant] == 0:
+                queue.append(dependant)
+
+    if initialized < len(degrees):
+        # Happens if the ModuleInterface for dependency doesn't support import mode
+        # The case when there is circular imports should already be covered by python loader
+        raise ImportError("Not all modules have been initialized. Check if your import modes are correct")
+
+
+def _initialize_module(module):
+    module()  # call a constructor for initialization
+
+
+def _check_loading_correct(in_config: Set[str], modes: List[ImportMode]):
+    """
+    To avoid unintentional loading (without ModuleInterface) we can check all loaded modules.
+    Unintentional/incorrect loading might happen if module A has a direct reference to module B without declaring it
+    in ModuleInterface. Doing so, this might lead to a problem when a module interface require to load something during
+    initialization. But since ModuleInterface is not initializing properly (using depends_on)
+    some functionality may work wrongly.
+    """
+    expected_load = set()
+    for module in _all_modules():
+        if module.name() in in_config:
+            expected_load.add(module)
+
+    to_add = list(expected_load)
+    while to_add:
+        new_to_add = []
+        while to_add:
+            module = to_add.pop()
+            for dependency in module.depends_on():
+                if dependency not in expected_load:
```

Comment:
I believe here we should also check if the dependency (the module) supports the input ImportModes, and only in that case add it to the set/list.

More generic question which came up for me due to this case: is it allowed (and fine) to declare module dependencies with different ImportMode support?